### PR TITLE
refactor: authorization checks in Tasklist

### LINKED ISSUE #21818: C8 Single app - Cannot Assign/Complete Zeebe user tasks using Tasklist
### To Reproduce

1. Run Tasklist + Zeebe broker as C8 single app (same when Operate is also running in the single app)
2. Configure Tasklist to use Zeebe instance running in the same app
3. Deploy and start a process containing Zeebe user task
4. Assign the Zeebe user task using Tasklist UI
5. Internal error is returned (authentication error returned when Tasklist calls Zeebe)

### Additional context

This is discussed in https://camunda.slack.com/archives/C06HTSPD5AP/p1724680654652539

It is caused by:
- Tasklist applies security filters on path `/v*/*` see [here](https://github.com/camunda/camunda/blob/9524ebc02e639f335f73daf85b770b86067c4d0f/tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/BaseWebConfigurer.java#L95).
- Tasklist calls `/v2/users-tasks` Zeebe REST endpoints to assign/complete user tasks
- When we run Tasklist alongside Zeebe, the Tasklist Spring Security package is (undesirably) applied to Zeebe's REST endpoints, in addition to the existing Zeebe authentication interceptors. While the call from Tasklist to Zeebe is authenticated, it is being rejected by Tasklist's Spring Security checks.

**Long term fix**: UI should directly invoke `/v2/users-tasks` endpoints instead of calling Tasklist /v1 endpoints.

---------------------------------------------------------------------------------------------

### Definition of Ready - Checklist

- [ ] The bug has been reproduced by the assignee in the environment compatible with the provided one; otherwise, the issue is closed with a comment
- [ ] The issue has a meaningful title, description, and testable acceptance criteria
- [ ] Necessary screenshots, screen recordings, or files are attached to the bug report

For UI changes required to solve the bug:

- [ ] Design input has been collected by the assignee

### After solving the issue - Answer and document in the comments the questions below

- [ ] What was the root cause for this bug?
- [ ] Is there anything we should do for the error to not happen anymore?

### LINKED ISSUE #24808: Unable to Assign Zeebe User Task in C8Run 8.7.0-alpha1 & 8.6.5
**Describe the issue:**

https://jira.camunda.com/browse/SUPPORT-25953

When using C8Run with `8.7.0-alpha1` and `8.6.5` , I am unable to assign a Zeebe User Task. 

**Actual behavior:**

When attempting to assign a Zeebe User Task, the task does not get assigned as expected.

**Expected behavior:**

I would expect to be able to assign a Zeebe User Task without any issues.

**How to reproduce:**

1. Set up C8Run with version `8.7.0-alpha1` or `8.6.5` .
2. Deploy a diagram with a Zeebe User Task
3. Attempt to assign a Zeebe User Task in the UI.
4. Observe that the task is not assigned and an error appears in the UI (see screenshot)
Screenshot 2024-11-15 at 17 33 31

*Note:* The issue occurs across all operating systems.

**Logs:**

I have attached logs and a screenshot of the UI behavior for reference.
[camunda.log](https://github.com/user-attachments/files/17799747/camunda.log)

**Environment:**

**Please note: Without the following info, it's hard to resolve the issue and probably it will be closed.**

- **Platform:** Mac, Windows, Linux
- **Architecture:** x86, ARM
- **Version:** 8.7.0-alpha1 and 8.6.5

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

* Replaces the `hasPermission()` checks with the new authorization concept, i.e., checking if a user/client is allowed to complete, assign, and unassign a task and creating an instance.
* Implements two modes:
  * New Architecture: Instead of communicating with the Gateway, it directly executes broker requests. In that case, the engine will perform the authorization checks.
  * Old Architecture (compatibility): This mode ensures that the old backend makes requests to the Gateway to ensure a running dev environment and old test infrastructure. However, the old backend does the authorization checks using the `AuthorizationChecker`.
* Provides integration tests.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

related to #24665

### CODE REVIEW

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/service/ProcessService.java [hunk 1]
```diff
@@ -83,53 +80,42 @@ public ProcessInstanceDTO startProcessInstance(
...
+      final String processDefinitionKey,
+      final List<VariableInputDTO> variables,
+      final String tenantId,
+      final boolean executeWithAuthentication) {
 
     final boolean isMultiTenancyEnabled = tenantService.isMultiTenancyEnabled();
 
     if (isMultiTenancyEnabled && !tenantService.getAuthenticatedTenants().contains(tenantId)) {
       throw new InvalidRequestException("Invalid tenant.");
     }
```
Comment (remcowesterhoud) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/service/ProcessService.java:96]: ❓ If no one is authenticated, how will we get the authenticated tenants?

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java [hunk 2]
```diff
@@ -490,6 +490,18 @@ public static Authentication getAuthentication() {
...
+  public static Authentication getAnonymousAuthentication() {
+    return new Builder()
+        .token(
+            Authorization.jwtEncoder()
+                .withIssuer(JwtAuthorizationBuilder.DEFAULT_ISSUER)
+                .withAudience(JwtAuthorizationBuilder.DEFAULT_AUDIENCE)
+                .withSubject(JwtAuthorizationBuilder.DEFAULT_SUBJECT)
+                .withClaim(Authorization.AUTHORIZED_ANONYMOUS_USER, true)
+                .build())
+        .build();
```
Comment (remcowesterhoud) [inline zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java:502]: 💭 When we implemented https://github.com/camunda/camunda/issues/23093 we don't need this JWT anymore. We could just have an `ANONYMOUS` [AuthDataFormat](https://github.com/camunda/camunda/blob/main/zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/encoding/AuthInfo.java#L97) which we could check for in Zeebe. That makes our life a bit easier.

**Review by remcowesterhoud — DISMISSED** [review]
@romansmirnov This PR definitely also needs someone from the HTO team to look at it as it's very much tasklist related.

The anonymous auth on the command seems alright. What's missing is that Zeebe actually has to take action to approve authorization in the [AuthCheckBehavior](https://github.com/camunda/camunda/blob/main/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationCheckBehavior.java#L64). Otherwise, it'll still get rejected.

File: qa/util/pom.xml [hunk 3]
```diff
@@ -29,6 +29,17 @@
       <artifactId>elasticsearch</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>org.awaitility</groupId>
+      <artifactId>awaitility</artifactId>
+      <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (romansmirnov) [reply qa/util/pom.xml:35]: I know, but maven and my IDE were complaining about it. I was only able to resolve this by adding the compile scope explicitly.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/service/ProcessService.java (hunk 1 above)
Comment (romansmirnov) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/service/ProcessService.java:96]: If no authentication is provided, then no tenant check will happen. This is handled in the tenant service.

File: zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java (hunk 2 above)
Comment (romansmirnov) [reply zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java:502]: That was actually my first implementation. Will #23093 be implemented with 8.7?

Comment (remcowesterhoud) [reply zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java:508]: I know it's on @nicpuppa his radar so it likely will be. But he's working on some more urgent issues atm.

Comment (nicpuppa) [reply zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java:508]: Yes, as @remcowesterhoud said I started working on [it](https://github.com/camunda/camunda/pull/25655). At the moment is a WIP

File: tasklist/common/src/main/java/io/camunda/tasklist/util/ConditionalOnTasklistCompatibility.java
Comment (houssain-barouni) [inline tasklist/common/src/main/java/io/camunda/tasklist/util/ConditionalOnTasklistCompatibility.java:1]: 👍

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/TaskController.java [hunk 4]
```diff
@@ -431,13 +429,14 @@ private boolean isUserRestrictionEnabled() {
...
-  @PreAuthorize("hasPermission('write')")
   @PostMapping("{taskId}/variables")
   public ResponseEntity<Void> saveDraftTaskVariables(
       @PathVariable @Parameter(description = "The ID of the task.", required = true)
           final String taskId,
       @RequestBody final SaveVariablesRequest saveVariablesRequest) {
-    if (!isUserRestrictionEnabled() || hasAccessToTask(getTaskSupplier(taskId))) {
+    final var taskSupplier = getTaskSupplier(taskId);
+    if (permissionServices.hasPermissionToUpdateUserTask(TaskDTO.toTaskEntity(taskSupplier.get()))
+        && (!isUserRestrictionEnabled() || hasAccessToTask(getTaskSupplier(taskId)))) {
```
Comment (houssain-barouni) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/TaskController.java:439]:
```suggestion
        && (!isUserRestrictionEnabled() || hasAccessToTask(taskSupplier))) {
```

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/CustomMethodSecurityExpressionHandler.java
Comment (houssain-barouni) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/CustomMethodSecurityExpressionHandler.java:1]: 👍

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  private boolean isAuthorizationDisabled() {
+    return !securityConfiguration.getAuthorizations().isEnabled();
+  }
+
+  private boolean isWithoutAuthenticatedUserKey(final Authentication authentication) {
+    // when the provided authentication does not contain a user key,
+    // then the authorization check cannot be performed.
+    // the authorization key is only provided when running with profile "auth-basic",
```
Comment (houssain-barouni) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java:76]: I suppose the authorization "auth-oidc" (based on mapping rules) should be implemented afterward

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      final Authentication authentication,
+      final Authorization authorization) {
+
+    if (isAuthorizationCheckDisabled(authentication)) {
+      return true;
+    }
+
+    final var securityContext =
+        securityContextProvider.provideSecurityContext(authentication, authorization);
+    return authorizationChecker.isAuthorized(resourceId, securityContext);
```
Comment (houssain-barouni) [inline tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java:62]:
```suggestion
    return securityContextProvider.isAuthorized(resourceId, authentication, authorization);
```
This way no need to autowire `SecurityConfiguration` and `AuthorizationChecker` and do checks if the feature is enabled

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/it/util/AuthorizationsUtil.java [hunk 7]
```diff
@@ -122,50 +127,53 @@ public boolean shouldRetryRequest(final StatusCode statusCode) {
...
-                  new URI(
-                      "http://%s/%s/_count/"
-                          .formatted(
-                              elasticsearchUrl, new UserIndex("", true).getFullQualifiedName())))
-              .header("Content-Type", "application/json")
-              .build();
-    } catch (final URISyntaxException e) {
-      throw new RuntimeException(e);
-    }
+    final var userQuery = UserQuery.of(b -> b.filter(f -> f.username(username)));
```
Comment (houssain-barouni) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/it/util/AuthorizationsUtil.java:130]: 👍

**Review by houssain-barouni — COMMENTED** [review]
This is excellent work—kudos @romansmirnov ! 🚀
I’ve added a few minor comments and questions.
We also need to update the [docker-compose](https://github.com/camunda/camunda/blob/main/tasklist/config/docker-compose.yml) file to set Zeebe compatibility enabled to true.
The failing OpenSearch IT has already been fixed in main.
I do wonder if this approach could be generalized for Operate, though implementing it there would likely be more challenging.

File: dist/src/main/java/io/camunda/tasklist/TasklistModuleConfiguration.java [hunk 8]
```diff
@@ -46,6 +47,7 @@
     // versions of importer
     nameGenerator = FullyQualifiedAnnotationBeanNameGenerator.class)
 @Profile("tasklist")
+@Import(TasklistPermissionServices.class)
```
[the remark this answers is not in the document]
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/tasklist/TasklistModuleConfiguration.java:50]: Yep, as long as the security package is excluded, it must explicitly be imported. I tested it in Standalone mode, and so far, it hasn't caused any issues.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java (hunk 5 above)
Comment (romansmirnov) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java:76]: Exactly.

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java (hunk 6 above)
Comment (romansmirnov) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/security/permission/TasklistPermissionServices.java:62]: Good Point, I noticed it when working on the second PR https://github.com/camunda/camunda/pull/26088 - there I applied that change https://github.com/camunda/camunda/pull/26088/files#diff-c96430bc8895499d0aabe0edf7a2f1f1c93a941a2951e3c4b388c236b099d64aR80

**Comment by romansmirnov** [comment]
> We also need to update the [docker-compose](https://github.com/camunda/camunda/blob/main/tasklist/config/docker-compose.yml) file to set Zeebe compatibility enabled to true.

@houssain-barouni, thank you for your review. To make sure, it must be set at for `tasklist-e2e` and `tasklist-sso`, right?

**Comment by houssain-barouni** [comment]
> > We also need to update the [docker-compose](https://github.com/camunda/camunda/blob/main/tasklist/config/docker-compose.yml) file to set Zeebe compatibility enabled to true.
> 
> @houssain-barouni, thank you for your review. To make sure, it must be set at for `tasklist-e2e` and `tasklist-sso`, right?

Also `tasklist` which we use the most

File: tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/TaskController.java (hunk 4 above)
Comment (romansmirnov) [reply tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/TaskController.java]: Fixed with https://github.com/camunda/camunda/pull/25799/commits/0bf0bafa695b38f34f890153e2f10c93d681f248

File: qa/integration-tests/src/test/java/io/camunda/it/tasklist/TasklistAssignUserTaskAuthorizationIT.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+  @AutoCloseResource private static ZeebeClient adminCamundaClient;
+  @AutoCloseResource private static TestRestTasklistClient tasklistRestClient;
+
+  private long processInstanceKey;
+  private long processInstanceKeyWithJobBasedUserTask;
+  private long userTaskKey;
+  private long userTaskKeyWithJobBasedUserTask;
+
+  @TestZeebe
+  private TestStandaloneCamunda broker =
```
[the remark this answers is not in the document]
Comment (romansmirnov) [reply qa/integration-tests/src/test/java/io/camunda/it/tasklist/TasklistAssignUserTaskAuthorizationIT.java]: Adjusted with https://github.com/camunda/camunda/pull/25799/commits/144598c9e4cac03a2992ef4de3f2662bbdb7d32c

**Review by houssain-barouni — APPROVED** [review]
LGTM!

**Pull request merged**